scylla/src/transport/node.rs

```diff
@@ -0,0 +1,173 @@
+
+        let shard_info_sender = Arc::new(std::sync::Mutex::new(shard_info_sender));
+
+        let connections = Arc::new(AsyncRwLock::new(NodeConnections::Single(
+            ConnectionKeeper::new(address, compression, None, Some(shard_info_sender.clone())),
+        )));
+
+        let worker = NodeWorker {
+            node_conns: connections.clone(),
+            node_addr: address,
+            compression,
+            shard_info_sender,
+            shard_info_receiver,
+        };
+
+        let (fut, worker_handle) = worker.work().remote_handle();
+        tokio::spawn(fut);
+
+        Node {
+            address,
+            datacenter,
+            rack,
+            connections,
+            _worker_handle: worker_handle,
+        }
+    }
+
+    /// Get connection which should be used to connect using given token
+    pub async fn connection_for_token(&self, token: Token) -> Result<Arc<Connection>, QueryError> {
+        let connections = &*self.connections.read().await;
+
+        match connections {
+            NodeConnections::Single(conn_keeper) => conn_keeper.get_connection().await,
+            NodeConnections::Sharded {
+                shard_info,
+                shard_conns,
+            } => {
+                let shard: Shard = shard_info.shard_of(token);
+                // It's guaranteed by NodeWorker that for each shard < shard_info.nr_shards
+                // there is an entry for this shard in connections map
+                shard_conns.get(&shard).unwrap().get_connection().await
+            }
+        }
+    }
+}
+
+impl NodeWorker {
+    pub async fn work(mut self) {
+        let mut cur_shard_info: Option<ShardInfo> = self.shard_info_receiver.borrow().clone();
+
+        loop {
+            // Wait for current shard_info to change
+            // We own one sending end of this channel so it can't return None
+            self.shard_info_receiver
+                .changed()
+                .await
+                .expect("Bug in NodeWorker::work");
+            let new_shard_info: Option<ShardInfo> = self.shard_info_receiver.borrow().clone();
+
+            // See if the node has resharded
+            match (&cur_shard_info, &new_shard_info) {
+                (Some(cur), Some(new)) => {
+                    if cur.nr_shards == new.nr_shards && cur.msb_ignore == new.msb_ignore {
+                        // Nothing chaged, go back to waiting for a change
+                        continue;
+                    }
+                }
+                (None, None) => continue, // Nothing chaged, go back to waiting for a change
+                _ => {}
+            }
+
+            cur_shard_info = new_shard_info;
+
+            // We received updated node ShardInfo
+            // Create new node connections. It will happen rarely so we can probably afford it
+            // TODO: Maybe save some connections instead of recreating?
```

Comment:
I don't know if I follow this comment correctly, but if we already received ShardInfo from a node, we would never receive another one unless the node was restarted first, which also implies that all of its connections are broken. Changing the number of shards in Scylla is only available upon restart.

Reply:
The idea was that after Node has resharded its connection will die and it's `ConnectionKeeper` will create a new `Connection` which will send new `ShardInfo` to `NodeWorker`. Then `NodeWorker` detects that there was a change and creates new connections to correct shards.

Reply:
Makes sense. This TODO could be removed though - since all connections are reset, you won't be able to save any, so it only makes sense to create all of them from scratch.

Reply:
They are reset but their `ConnectionKeeper` will immediately reconnect so it's a bit wasteful to discard it's work altogether.